sympy/combinatorics/perm_groups.py

```diff
@@ -1702,6 +1705,79 @@ def is_abelian(self):
                     return False
         return True
 
+    def abelian_invariants(self):
+        """
+        Returns the abelian invariants for the given group.
+        Let ``G`` be a nontrivial finite abelian group. Then G is isomorphic to
+        the direct product of finitely many nontrivial cyclic groups of
+        prime-power order.
+
+        The prime-powers that occur as the orders of the factors are uniquely
+        determined by G. More precisely, the primes that occur in the orders of the
+        factors in any such decomposition of ``G`` are exactly the primes that divide
+        ``|G|`` and for any such prime ``p``, if the orders of the factors that are
+        p-groups in one such decomposition of ``G`` are ``p^{t_1} >= p^{t_2} >= ... p^{t_r}``,
+        then the orders of the factors that are p-groups in any such decomposition of ``G``
+        are ``p^{t_1} >= p^{t_2} >= ... p^{t_r}``.
+
+        The uniquely determined integers ``p^{t_1} >= p^{t_2} >= ... p^{t_r}``, taken
+        for all primes that divide ``|G|`` are called the invariants of the nontrivial
+        group ``G`` as suggested in ([14], p. 542).
+
+        Notes
+        =====
+
+        We adopt the convention that the invariants of a trivial group are [].
+
+        Examples
+        ========
+
+        >>> from sympy.combinatorics import Permutation
+        >>> from sympy.combinatorics.perm_groups import PermutationGroup
+        >>> a = Permutation([0, 2, 1])
+        >>> b = Permutation([1, 0, 2])
+        >>> G = PermutationGroup([a, b])
+        >>> G.abelian_invariants()
+        [2]
+        >>> from sympy.combinatorics.named_groups import CyclicGroup
+        >>> G = CyclicGroup(7)
+        >>> G.abelian_invariants()
+        [7]
+
+        """
+        if self.is_trivial:
+            return []
+        gns = self.generators
+        inv = []
+        der = self.derived_subgroup()
+        G = self
+
+        for p in primefactors(G.order()):
+            ranks = []
+            r = 2
+            while r > 1:
+                H = der
+                l = []
+                for g in gns:
+                    elm = g**p
+                    if not H.contains(elm):
+                        l.append(elm)
+                H = PermutationGroup(H.generators + l)
+                r = G.order()//H.order()
+                G = H
+                gns = G.generators
```

Comment:
`gns` now contains also the generators of `der`. Would it suffice to take only those that are in the list `l` of powers?

Reply:
I've checked and yes it's consistent with the previous result.

Reply:
In that case, I would suggest that the creation of `l` above be changed. Maybe something like this could work:

    pows = [g**p for g in gns]
    gns = [pow in pows if not H.contains(pow)]

It should be possible to use the name `gns` (or maybe `gens`) at this point already.

Reply:
I think the above `gns` should be `gns = [pow for pow in pows if not H.contains(pow)]`and in that case does it really make a difference, for me both(current and the above mentioned changes) looks same.

Reply:
Yes, of course.

Reply:
After applying above changes one of the test-case seems to fail i.e the result is not consistent with GAP.

```
>>> G = AbelianGroup(2,3,4)
>>> assert G.abelian_invariants() == [2, 3, 4]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError
>>> G.abelian_invariants()
[2, 4]
```

Reply:
Perhaps this `H = PermutationGroup(H.generators + l)` should have `dir.generators` instead of the changing `H.generators`.

Reply:
`H = PermutationGroup(H.generators + l)` works fine the above failure is encountered after below changes:
```
pows = [g**p for g in gns]
gns = [pow for pow in pows if not H.contains(pow)]

```
And the reason seems to be same that you have mentioned above after calculating `pows` we are updating `H = PermutationGroup(H.generators + pows)` where `H.generators` is always `der.generators`.

So, I think the current commit is okay!